File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
      ListNode* dummy1 = new ListNode(1);
      ListNode* cur = dummy1;
      ListNode* cur1 = list1;
      ListNode* cur2 = list2;

      while(cur1 && cur2){
        if(cur1->val <= cur2->val){
            cur->next = cur1;
            cur1 = cur1->next;
        }else{
            cur->next = cur2;
            cur2 = cur2->next;
        }
        cur = cur->next;
      }
      if(cur1){
        cur->next = cur1;
      }else{
        cur->next = cur2;
      }
      return dummy1->next;

    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse(node, end):
            prev = None
            cur = node
            while cur != end:
                curNext = cur.next
                cur.next = prev
                prev = cur
                cur = curNext
            return prev
        dummy = ListNode()
        dummy.next = head
        fast = dummy
        while True:
            slow = fast
            for _ in range(k):
                fast = fast.next
                if fast is None:
                    return dummy.next
            nextGroupHead = fast.next
            curGroupEnd = slow.next
            curGroupHead = reverse(curGroupEnd, nextGroupHead)
            slow.next = curGroupHead
            curGroupEnd.next = nextGroupHead
            fast = curGroupEnd
        

